src/add.rs

```diff
@@ -209,10 +210,33 @@ enum ChezmoiState {
     },
 }
 
+// Source - https://stackoverflow.com/questions/26076005/how-can-i-list-files-of-a-directory-in-rust
+fn recurse_files(path: &Utf8Path) -> anyhow::Result<Vec<Utf8PathBuf>> {
+    let mut buf = vec![];
+    let entries = Utf8Path::read_dir_utf8(path)?;
+
+    for entry in entries {
+        let entry = entry?;
+        let meta = entry.metadata()?;
+
+        if meta.is_dir() {
+            let mut subdir = recurse_files(entry.path())?;
+            buf.append(&mut subdir);
```

Comment:
This code is not on what I consider the "hot path" so the moving of strings here is not a big issue. But a way to avoid that (and reduce the number of allocations) would be to change the signature to

```rust
fn recurse_files(path: &Utf8Path, buf: &mut Vec<Utf8PathBuf>) -> anyhow::Result<()>
```

(For reference, the "hot path" is when executing in `chezmoi apply` and similar. The adding-path is not as performance critical. However, optimising code is still one of my favourite parts of software development, so I can't resist when I see the opportunity.)